### DISCUSSION TITLE
UserWarning: does not have profile information

### DISCUSSION THREAD

**Discussion by Jellun** [snakers4/silero-vad#379] [General]
Hi All,

The model works well when the following code is put in the Global scope and when there it's used in the main thread:
model, utils = torch.hub.load(repo_or_dir='snakers4/silero-vad',
                                                                    model='silero_vad',
                                                                    force_reload=False,
                                                                    trust_repo=True,
                                                                    onnx=USE_ONNX)
(get_speech_timestamps,
      save_audio,
      read_audio,
      VADIterator,
      collect_chunks) = utils

It starts to fail if multithreads are using the model.

If I put the above code inside a function that is used by multithreads, the model runs successfully, except, it gives the following warning message:
C:\Users\JYe\AppData\Local\Programs\Python\Python311\Lib\site-packages\torch\nn\modules\module.py:1501: UserWarning: operator () profile_node %669 : int[] = prim::profile_ivalue(%667)
 does not have profile information (Triggered internally at ..\third_party\nvfuser\csrc\graph_fuser.cpp:108.)
  return forward_call(*args, **kwargs)

Does anyone know how to get rid of this warning message properly, rather than disabling the warning messages by warnings.filterwarnings("ignore"). Note that I run the model on a Windows 10 machine.

**Comment by snakers4** [comment]
Hi,

The proper way to use the PyTorch model at least, is to:

- set 1 thread in PyTorch;
- have 1 instance of model per separate CPU thread;
- have some sort of bufffering / queue / messenger mechanism to make sure that 2 different clients do not use the same model in the same thread at the same time;

PyTorch models behave weirdly, expecially under load, when you have several threads working with the same model.

**Reply by Jellun** [reply]
![image](https://github.com/snakers4/silero-vad/assets/20332616/e3f71974-3e51-496d-a1ca-399f937ea941)
The same warning message occurs even when there is only one child thread processing one single file.

![image](https://github.com/snakers4/silero-vad/assets/20332616/da03bdb4-477f-4f7f-9a20-bc34dc00dd0e)
Here, def process_audio(fileName) function is running in a child thread.
As long as the code initializing silero-vad is moved outside of process_audio function, there is no warning message, however, by doing that, it will generate fatal errors when there are more than one threads processing more than one audio files.

I opted for initializing silero-vad model inside a child thread, so it can deal with multiple audio files simultaneously. It actually works - all multiple files are processed successfully to the end and accurate results are produced. It would just be really nice if one can help get rid of that warning messages.

**Reply by snakers4** [reply]
> I opted for initializing silero-vad model inside a child thread, so it can deal with multiple audio files simultaneously. It actually works - all multiple files are processed successfully to the end and accurate results are produced. It would just be really nice if one can help get rid of that warning messages.

At the same time, one model should process one separate audio stream, otherwise strange errors may occur.
If you need to process several streams with one model, model can support batching, bit this is also tricky, and also discouraged.
